File: courses/lectures/inv-calc.ml
type prop =
  | And of prop * prop
  | Impl of prop * prop
  | Or of prop * prop
  | Top
  | Bot
  | Atom of string

let rec equal gamma gamma' =
  match gamma with
  | a::tl -> List.exists (fun a' -> a = a') gamma' && equal tl gamma'
  | [] -> true

let loop memo (gamma, prop) =
  List.exists (fun (gamma', prop') -> prop = prop' && equal gamma gamma' && equal gamma' gamma) memo

let rec invR memo gamma omega prop =  (* gamma ; omega --R--> prop *)
  match prop with
  | And(a, b) -> invR memo gamma omega a && invR memo gamma omega b
  | Impl(a, b) -> invR memo gamma (a::omega) b
  | Top -> true
  | x -> invL memo gamma omega x

and invL memo gamma omega c =     (* gamma ; omega  --L--> c *)
  match omega with
  | And(a, b)::omega' -> invL memo gamma (a::b::omega') c
  | Or(a, b)::omega' -> invL memo gamma (a::omega') c && invL memo gamma (b::omega') c
  | Top::omega' -> invL memo gamma omega' c
  | Bot::_omega' -> true
  | x::omega' -> invL memo (x::gamma) omega' c
  | [] -> choice memo gamma c

and choice memo gamma prop =     (* gamma --C--> prop *)
  if loop memo (gamma, prop) then false
  else choiceR ((gamma, prop)::memo) gamma prop || choiceL ((gamma, prop)::memo) gamma prop

and choiceR memo gamma prop =
  match prop with
  | Or(a, b) -> invR memo gamma [] a || invR memo gamma [] b
  | Atom(p) -> List.exists (fun q -> q = Atom(p)) gamma
  | _ -> false    (* dead code *)

and choiceL memo gamma prop =
  List.exists
    (fun p -> match p with
              | Impl(a, b) -> invR memo gamma [] a && invL memo gamma [b] prop
              | Atom(_p) -> false
              | _ -> false
    )
  gamma

File: courses/lectures/inv_calc.pl
exists(A, [A | Gamma]).

exists(A, [B | Gamma]) :- exists(A, Gamma).

inv_prove(Gamma, A) :- invR([], Gamma, A).

invR(Gamma, Omega, and(A, B)) :-
    invR(Gamma, Omega, A),
    invR(Gamma, Omega, B).

invR(Gamma, Omega, or(A, B)) :-
    invL(Gamma, Omega, or(A, B)).

invR(Gamma, Omega, impl(A, B)) :-
    invR(Gamma, [A | Omega], B).

invR(Gamma, Omega, atom(P)) :- invL(Gamma, Omega, atom(P)).

invL(Gamma, [and(A, B) | Omega], C) :-
    invL(Gamma, [A, B | Omega], C).

invL(Gamma, [or(A, B) | Omega], C) :-
    invL(Gamma, [A | Omega], C),
    invL(Gamma, [B | Omega], C).

invL(Gamma, [impl(A, B) | Omega], C) :-
    invL([impl(A, B) | Gamma], Omega, C).

invL(Gamma, [atom(P) | Omega], C) :-
    invL([atom(P) | Gamma], Omega, C).

invL(Gamma, [], C) :-
    choice(Gamma, C).

choice(Gamma, C) :- choiceR(Gamma, C).

choice(Gamma, C) :- choiceL(Gamma, C).

choiceR(Gamma, or(A, B)) :- invR(Gamma, [], A).

choiceR(Gamma, or(A, B)) :- invR(Gamma, [], B).

choiceR(Gamma, atom(P)) :- exists(atom(P), Gamma).

choiceL(Gamma, C) :-
    exists(impl(A, B), Gamma),
    invR(Gamma, [], A),
    invL(Gamma, [B], C).